include/graphblas/bsp/collectives.hpp

```diff
@@ -69,6 +70,55 @@
 
 namespace grb {
 
+	namespace {
+
+		/**
+		 * This is a reducer function of the signature specified by lpf_reducer_t.
+		 */
+		template< typename OP >
+		void generic_reducer( size_t n, const void * array_p, void * value_p ) {
+			typedef typename OP::D1 lhs_type;
+			typedef typename OP::D2 rhs_type;
+			typedef typename OP::D3 out_type;
+			static_assert(
+					std::is_same< rhs_type, out_type >::value,
+					"A generic operator-only reducer from an array into a scalar requires the "
+					"RHS and output types to be the same"
+				);
+			assert( array_p != value_p );
+
+			const lhs_type * const __restrict__ array =
+				static_cast< const lhs_type * >( array_p );
+			out_type * const __restrict__ value =
+				static_cast< out_type * >( value_p );
+
+			lhs_type left_buffer[ OP::blocksize ];
```

Comment:
this array is allocated on the stack (and should assume `OP::blocksize` is `constexpr`); for small `OP::blocksize` this is not a problem, but for large ones it may be (e.g., SVE vector size of 2K, with 8K stack size limit typically enforced on Linux). This is not a concern under the **guarantee** that the compiler is able to vectorize the following loop and hence eliminate this temporary buffer.

Reply:
This trick is used throughout ALP to guide the compiler to vectorise properly. The (indeed `constexpr`) blocksize reflects whatever fits in the configured SIMD length, and the compiler replaces this buffer with a vector register.